content/docs/context.md

```diff
@@ -4,220 +4,110 @@ title: Context
+```js
+<Consumer>
+  {value => /* render something based on the context value */}
+</Consumer>
+```
 
-> Note:
->
-> As of React 16, `componentDidUpdate` no longer receives `prevContext`.
+A React component that subscribes to context changes.
 
-## Referencing Context in Stateless Functional Components
+Requires a [function as a child](/docs/render-props.html#using-props-other-than-render). The function receives the current context value and returns a React node. All consumers are re-rendered whenever the Provider value changes. Changes are determined by comparing the new and old values using [`Object.is`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/is).
 
-Stateless functional components are also able to reference `context` if `contextTypes` is defined as a property of the function. The following code shows a `Button` component written as a stateless functional component.
+> Note
+> 
+> For more information about this pattern, see [render props](/docs/render-props.html).
 
-```javascript
-import PropTypes from 'prop-types';
+## Examples
 
-const Button = ({children}, context) =>
-  <button style={{background: context.color}}>
-    {children}
-  </button>;
+### Dynamic Context
 
-Button.contextTypes = {color: PropTypes.string};
-```
+A more complex example with dynamic values for the theme:
 
-## Updating Context
+**theme-context.js**
+`embed:context/theme-detailed-theme-context.js`
 
-Don't do it.
+**themed-button.js**
+`embed:context/theme-detailed-themed-button.js`
 
-React has an API to update context, but it is fundamentally broken and you should not use it.
+**app.js**
+`embed:context/theme-detailed-app.js`
 
-The `getChildContext` function will be called when the state or props changes. In order to update data in the context, trigger a local state update with `this.setState`. This will trigger a new context and changes will be received by the children.
+## Consuming Multiple Contexts
 
-```javascript
-import PropTypes from 'prop-types';
+`embed:context/multiple-contexts.js`
 
-class MediaQuery extends React.Component {
-  constructor(props) {
-    super(props);
-    this.state = {type:'desktop'};
-  }
+## Accessing Context in Lifecycle Methods
 
-  getChildContext() {
-    return {type: this.state.type};
-  }
+`embed:context/lifecycles.js`
 
-  componentDidMount() {
-    const checkMediaQuery = () => {
-      const type = window.matchMedia("(min-width: 1025px)").matches ? 'desktop' : 'mobile';
-      if (type !== this.state.type) {
-        this.setState({type});
-      }
-    };
+## Forwarding Refs to Context Consumers
 
-    window.addEventListener('resize', checkMediaQuery);
-    checkMediaQuery();
-  }
+**fancy-button.js**
+`embed:context/forwarding-refs-fancy-button.js`
 
-  render() {
-    return this.props.children;
-  }
-}
+**app.js**
+`embed:context/forwarding-refs-app.js`
 
-MediaQuery.childContextTypes = {
-  type: PropTypes.string
-};
-```
+## Legacy API
```

Comment:
Before "Legacy API", we need to add a "Caveats" section.

There, we should mention that pattern like `<Provider value={{ something: 'something' }}>` re-renders all consumers every time the provider parent re-renders because the `value` object is referentially different every time. If you didn't intend this, put the `value` into the parent component's `state` so that it doesn't change unintentionally between re-renders.

Reply:
Would it make sense to put this in the Consumer API section?

Reply:
Maybe. What do you think? I generally prefer to separate "minimal necessary" information to use the API from all the details (that they user won't bump into later on), but maybe it's okay to do in one place. The main reason I suggested a separate section is because I wanted to have a "good" and "bad" example (even if they're one liners), and I felt it would they would steal the focus from the API reference if we put them there.

Reply:
OK, how about linking to Caveats/Gotchas after mentioning the Object.is comparison?

Reply:
Sounds goode